### ISSUE TITLE
Storybook 6.0 not picking up types from interfaces

### ISSUE THREAD

**Issue by chadlavi-casebook** [storybookjs/storybook#12558]
**Describe the bug**
I'm trying out storybook with an inner-source typescript react library I maintain at work (unfortunately, it is *not* public, but I think I can provide some anonymized examples).

Funtionally, storybook is working great, but it doesn't pick up the types for documentation. My "Docs" tab is the same thing the user who filed #12212 is seeing, just a recapitulation of the "Canvas" tab.

**Screenshots**
<img width="692" alt="Screen Shot 2020-09-23 at 5 00 09 PM" src="https://user-images.githubusercontent.com/45078010/94069280-4d41e880-fdbe-11ea-89b3-f3744a8b3fa5.png">

<img width="689" alt="Screen Shot 2020-09-23 at 5 00 19 PM" src="https://user-images.githubusercontent.com/45078010/94069292-52069c80-fdbe-11ea-8819-b667243c5c2b.png">

**Code snippets**
my `.storybook/main.js`: 

```js
module.exports = {
  "stories": [
    "../src/**/*.stories.mdx",
    "../src/**/*.stories.@(js|jsx|ts|tsx)"
  ],
  "addons": [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    "@storybook/preset-create-react-app"
  ],
  "typescript": {
    check: false,
    checkOptions: {},
    reactDocgen: 'react-docgen-typescript',
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      propFilter: (prop) => (prop.parent ? !/node_modules/.test(prop.parent.fileName) : true),
    },
  },
}
```

my `Button.stories.tsx`: 

```tsx
import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Button, ButtonProps } from 'myLibraryNameHere';

export default {
  title: 'Example/Button',
  component: Button,
  argTypes: {
    variant: {
      control: {
        type: 'select',
        options: ['destructive', 'outline', 'primary', 'standard', 'workflowAction']
      }
    },
    compact: {
      control: 'boolean',
    }
  },
  args: {
    children: 'Button',
    variant: 'standard',
  },
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary',
};

export const Standard = Template.bind({});
Standard.args = {
};

export const Compact = Template.bind({});
Compact.args = {
  compact: true,
};
```

And here's what the `Button.tsx` file in my library looks like (well, a very simplified view of its structure anyway): 

```tsx
import MuiButton from '@material-ui/core/Button'
import React, {forwardRef} from 'react'

import withTheme from '../hoc/withTheme'

import {ButtonProps} from './Button.model'
import style from './Button.style'

export const Button = forwardRef<HTMLElement, ButtonProps>(({...props}: ButtonProps, ref: React.Ref<HTMLElement>) => (
  <MuiButton {...props}/>
))

/**
 * JSDOC COMMENT
 */
const MyCompanyButton = withTheme(Button, style)

MyCompanyButton.displayName = 'Button'

MyCompanyButton.defaultProps = {
  component: 'button',
  variant: 'standard',
}

export default MyCompanyButton

```

**System:**

Environment Info:

  System:
    OS: macOS 10.15.6
    CPU: (4) x64 Intel(R) Core(TM) i5-7360U CPU @ 2.30GHz
  Binaries:
    Node: 14.11.0 - /usr/local/bin/node
    Yarn: 1.22.5 - /usr/local/bin/yarn
    npm: 6.14.8 - /usr/local/bin/npm
  Browsers:
    Chrome: 85.0.4183.121
    Firefox: 79.0
    Safari: 14.0
  npmPackages:
    @storybook/addon-actions: ^6.0.21 => 6.0.21 
    @storybook/addon-essentials: ^6.0.21 => 6.0.21 
    @storybook/addon-links: ^6.0.21 => 6.0.21 
    @storybook/node-logger: ^6.0.21 => 6.0.21 
    @storybook/preset-create-react-app: ^3.1.4 => 3.1.4 
    @storybook/react: ^6.0.21 => 6.0.21 

**Additional context**
We're forwarding refs because otherwise they can't be passed properly to the MUI component. Is that what's causing the issue here? or maybe the withTheme HOC?

We use `react-docgen-typescript` to generate type documentation without an issue, so I don't think this is an issue with docgen... any insight from you all would be appreciated. If I could fix this, I think I'd just move all my docs to a Storybook!

**Comment by chadlavi-casebook** [comment]
Additional note: the docs for default starter stories (header and page) that come with a new instance of storybook _are_ working just fine

**Comment by shilman** [comment]
I think this is the issue:

```js
import { Button, ButtonProps } from 'myLibraryNameHere';
```

Does it work if you replace it with:

```js
import { Button, ButtonProps } from '../../path/to/myLibrary';
```

Storybook doesn't currently extract types from `node_modules` https://github.com/storybookjs/storybook/issues/12185

**Comment by chadlavi-casebook** [comment]
oh @shilman ok, this must be it. I was trying to do this in a different project from where my library code is!

Thank you for the reply, I'll close this issue since it's a known issue/expected current behavior.

**Comment by chadlavi-casebook** [comment]
oops! Hey @shilman I'm back.

I've set up storybook in my lib repo, and it looks like the same problem occurs when I import from `../path/to/myLibrary`. None of the types are picked up.

**Comment by chadlavi-casebook** [comment]
I have this in the story file: 

```tsx
// Button.stories.tsx

import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import { Button, ButtonProps } from '../components';

export default {
  title: 'Example/Button',
  component: Button,
  args: {
    children: "Button",
  }
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary'
};

export const Standard = Template.bind({});
Standard.args = {
};

export const Compact = Template.bind({});
Compact.args = {
  compact: true,
};
```

`Button` and `ButtonProps` are named exports in `./src/components/index.ts`.

However, it _does_ work if I separately import the component from the file where it's defined and use _that_ for the component prop: 

```tsx
// Button.stories.tsx with separate import for component def

import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import { Button, ButtonProps } from '../components';

import { Button as B } from '../components/Button/Button'

export default {
  title: 'Example/Button',
  component: B,
  args: {
    children: "Button",
  }
} as Meta;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  variant: 'primary'
};

export const Standard = Template.bind({});
Standard.args = {
};

export const Compact = Template.bind({});
Compact.args = {
  compact: true,
};

```

inside `./src/components/Button/Button.tsx` I export an unstyled component as a named export, then default export a component styled with a `withTheme` HOC. Roughly like this: 

```tsx

// this is "B" in the Buttons.stories.tsx file
export const Button = forwardRef<HTMLElement, ButtonProps>((props: ButonProps, ref: Ref<HTMLElement>) => { ... })
const ThemedButton = withTheme(ButtonBase, style)

const MyButton = (props: ButtonProps) => <ThemedButton {...props} />

// This is used as the named export "Button" in src/components/index.ts
export default MyButton
```

Could it be the HOC that's throwing this off? The re-exporting a default export as a named export?

**Comment by shilman** [comment]
Yeah that sounds right. What happens if you set the `default.component` to be the inner Button instead of the HOC-wrapped one?

**Comment by chadlavi-casebook** [comment]
@shilman yep, that's what I'm doing with `import { Button as B } ...` in the stories.tsx file code above. The problem is that I don't want to export that, it should never be used in my library since it does not have styling applied.

the withTheme/withStyle stuff I'm using is out of Material UI. Have you all run into anyone else with similar issues with HOCs blocking react-docgen-typescript from picking up types? I took a look through their issue tracker and it looks like they (supposedly) do support reading the types from the return of an HOC.

**Comment by cieldon32** [comment]
I have the same issue.
if I change the types' export code to "export * from '.........'", there are no error. But I wouldn't like to export all. I need to export like "export {xxxxx, yyyyyyy} from '.......'".

**Comment by pdimova** [comment]
I have exactly the same issue as @cieldon32 

`export { A,B,C,D,} from './types';` throws an warning message in the console:

"export 'A' was not found in './types'
"export 'B' was not found in './types'
"export 'C' was not found in './types'

interestingly, no warning message for type D 🤷 

`export * from './types';` seems to cuase no troubles 

Is there something I can do about the following messages:
```
client.js:196
 ./src/../../emotion.ts 4:0-65 "export 'CSSObject' was not found in '@emotion/serialize'
./src/../../emotion.ts 4:0-65"export 'SerializedStyles' was not found in '@emotion/serialize'
```

**Comment by jeffersoneagley** [comment]
Same issue here. I'm extending an interface from an external library and none of the properties of the base type I extended show up.

**Comment by dbollinger** [comment]
This might not address all the different scenarios in this thread, but inferring properties from extended interfaces is possible by overriding default storybook config:

https://storybook.js.org/docs/ember/configure/typescript#overriding-the-configuration-to-infer-additional-props

**Comment by JuhG** [comment]
Exporting the bare component is not really a solution, it's missing functionality the HOC provides. Doesn't make much sense to show your component without styles for example. (any better solution since @chadlavi-casebook?)

`react-docgen-typescript` is able to get the props and documentation, even if we're only exporting the wrapped component. Also other tools like Styleguidist (using the same docgen library under the hood) is able to figure it out using the exact same component. (currently using both at the same time, this is the blocking issue migrating to SB)

Here's the [codesandbox](https://codesandbox.io/p/sandbox/condescending-knuth-gi31ll?selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A13%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A13%7D%5D&file=%2Fsrc%2Fscript.ts&workspace=%257B%2522activeFileId%2522%253A%2522cl9ix4l0b04443b6lekrho7xg%2522%252C%2522openFiles%2522%253A%255B%2522%252Ftsconfig.json%2522%252C%2522%252Fpackage.json%2522%252C%2522%252Fsrc%252Fwrapper.tsx%2522%252C%2522%252Fsrc%252FTwoWaySwitch.tsx%2522%252C%2522%252Fsrc%252Fscript.ts%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522sidekickItems%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522key%2522%253A%2522cl9ixdvq406k73b6lq8846eln%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D) with a HOC wrapper and no named export to prove `react-docgen-typescript` is able to figure the props out. (run start script)

If there's any way we can contribute to fix this issue, I'd be happy to help. Any pointers would be wonderful. Thanks!

Sorry for the duplicated comment on this closed issue: https://github.com/storybookjs/storybook/issues/12576#issuecomment-1287394740

**Comment by shilman** [comment]
@JuhG Thanks for your example. I looked in some detail and tried it out in Storybook. Here's what I found:

`TwoWaySwitch.tsx` declares two components, `TwoWaySwitch` and the wrapped default export. As far as I can tell, `react-docgen-typescript` correctly analyzes the types for `TwoWaySwitch`, but doesn't pick up anything for the default export.

Thus if you were to write your story like this (CSF3, can be backported to CSF2), you will see the desired prop table and the wrapped story:

```js
import Wrapper, { TwoWaySwitch } from './TwoWaySwitch';

export default {
  component: TwoWaySwitch,
  render: (args: any) => <Wrapper {...args} />,
};

export const Basic = {};
```

I suspect that styleguidist does a trick where they are doing a lookup based on the component's displayName. I think we could do something similar in Storybook, but it won't happen in 7.0. We plan to rewrite the props table code in a much better way in 7.x and could address this then. In the meantime I'd recommend the workaround above.

Please let me know if that all makes sense or if I'm missing something. Thanks!

**Comment by JuhG** [comment]
@shilman Thanks for the answer!

We'll do the workaround, but changing code based on developer tooling always feels wrong. :( It's good to hear, that there are plans on improving this! Thanks for your effort!

I assume there's no chance for a function similar to [propsParser](https://react-styleguidist.js.org/docs/configuration/#propsparser) to main.js which let's us override whatever SB detects? (because if I'm able to detect the props correctly in a node script, I could do that in main.js as well)

**Comment by shilman** [comment]
@JuhG The only change required is adding an extra named export to your component file. We do this all the time for writing tests -- this isn't too different. You also need to write your stories a little differently--which is unfortunate--but those stories aren't your production code.

I'm not going to touch this in the short term. We could add more configuration, but it would just lead to more stuff to back out of when we do the rewrite. I'd rather address it as part of the new setup. Thanks for your patience!

**Comment by Stan-Stani** [comment]
```tsx

interface IComponentProps {
  /** doc comment 1 */
  blah: string
  /** doc comment 2 */
  destructure: number
}

const Component: FunctionComponent<ComponentProps> = ({blah, destructure}): IComponentProps => {
  return "stuff"
}
```

Also worked in my case. Namely, the explicit definition of the props argument as `: IComponentProps` (even though it's already got a definition in the generic.

This way I don't have to export a named export.

Alternatively, defining the .displayName got the prop interface documentation into Storybook.

**Comment by valentinpalkovic** [comment]
Thanks for the report. This was on Storybook 6.0 with the CRA preset and the legacy `react-docgen-typescript` default. Since then the React docgen default flipped to `react-docgen` (Storybook 8), the `ArgsTable` block was replaced by `ArgTypes`/`Controls`, and CRA itself has been deprecated by React. The whole pipeline you're hitting is gone, so I'm closing this. Please open a new issue on Storybook 9/10 (with `@storybook/react-vite` or `@storybook/nextjs-vite`) if TS interfaces still don't show up.
